### PR TITLE
parser, planner: add query block level `no_decorrelate` hint

### LINKED ISSUE #37789: Support query block level `no_decorrelate()` hint
## Enhancement

Currently, the decorrelation in tidb is rule-based, not cost-based. But decorrelation doesn't guarantee to generate a better plan. So there are some cases that we need tidb not decorrelate a subquery but tidb directly decorrelates it, and generates a bad plan.

A typical case is like this: The outer query has a very small row count, and the subquery has an index on the correlated column. But there are some other operators (like aggregate) above the table in the subquery, which makes the IndexJoin inapplicable. If tidb decorrelates it, there would be a full scan in the subquery. However, we could have used the index to access only several rows from the table in the subquery.
Example:
```sql
create table t1(a int, b int);
create table t2(a int, b int, index ia(a));
explain select a > (select sum(b) from t2 where a = t1.b) from t1;
```
```
+------------------------------------+----------+-----------+---------------+-------------------------------------------------------------------------------------------+
| id                                 | estRows  | task      | access object | operator info                                                                             |
+------------------------------------+----------+-----------+---------------+-------------------------------------------------------------------------------------------+
| Projection_10                      | 10000.00 | root      |               | gt(cast(test.t1.a, decimal(20,0) BINARY), Column#10)->Column#11                           |
| └─HashJoin_11                      | 10000.00 | root      |               | left outer join, equal:[eq(test.t1.b, test.t2.a)]                                         |
|   ├─HashAgg_22(Build)              | 7992.00  | root      |               | group by:test.t2.a, funcs:sum(Column#13)->Column#10, funcs:firstrow(test.t2.a)->test.t2.a |
|   │ └─TableReader_23               | 7992.00  | root      |               | data:HashAgg_15                                                                           |
|   │   └─HashAgg_15                 | 7992.00  | cop[tikv] |               | group by:test.t2.a, funcs:sum(test.t2.b)->Column#13                                       |
|   │     └─Selection_21             | 9990.00  | cop[tikv] |               | not(isnull(test.t2.a))                                                                    |
|   │       └─TableFullScan_20       | 10000.00 | cop[tikv] | table:t2      | keep order:false, stats:pseudo                                                            |
|   └─TableReader_14(Probe)          | 10000.00 | root      |               | data:TableFullScan_13                                                                     |
|     └─TableFullScan_13             | 10000.00 | cop[tikv] | table:t1      | keep order:false, stats:pseudo                                                            |
+------------------------------------+----------+-----------+---------------+-------------------------------------------------------------------------------------------+
```
If the row count of t1 is very small and `a = t1.b` could filter most rows of data, row-by-row executed Apply will be much faster than the full scan + hash join strategy.

Though cost-based decorrelation is not easy for current tidb, we can provide a method to control this behavior manually.
Currently, there is only a global optimization rule blocklist to control this behavior:
https://docs.pingcap.com/tidb/dev/blocklist-control-plan#the-blocklist-of-optimization-rules-and-expression-pushdown
 It's very hard to use and not friendly.
To make it easier to control, we can implement a hint like other cost-based choices. To provide more flexibility, we can make this hint a query block level hint so that we can specifically choose which Apply we don't want to decorrelate.

### PR DESCRIPTION
**Pull request by time-and-fate**
### What problem does this PR solve?

Issue Number: close #37789

### What is changed and how it works?

Implement query block level `no_decorrelate()` hint.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/testdata/plan_suite_in.json
Comment (Reminiscent) [inline planner/core/testdata/plan_suite_in.json:1035]: Please add some test cases about `USE_TOJA` hint.
Besides, please use these hint in the same query block to see what happened if these hint are conflicts.

Comment (Reminiscent) [reply planner/core/testdata/plan_suite_in.json:1035]: Should we add these test cases about the TiFlash and partition tables(Both static and dynamic mode)?

File: planner/core/expression_rewriter.go
Comment (Reminiscent) [inline planner/core/expression_rewriter.go:321]: Please update the comments for this function.

File: planner/core/logical_plan_builder.go
Comment (Reminiscent) [inline planner/core/logical_plan_builder.go:3694]: Should `continue` here?

File: planner/core/expression_rewriter.go
Comment (Reminiscent) [inline planner/core/expression_rewriter.go:349]: It seems we always return `0` for the return value `hintFlags uint64`? Because we set `er.b.subQueryHintFlags = 0` in line335 and return always execute before the defer function in line337-341.

Comment (Reminiscent) [inline planner/core/expression_rewriter.go:349]: It seems we always return `0` for the return value `hintFlags uint64`? Because we set `er.b.subQueryHintFlags = 0` in line335 and return always execute before the defer function in line337-341.

File: planner/core/expression_rewriter.go
Comment (Reminiscent) [inline planner/core/expression_rewriter.go:540]: Should we reset the `noDecorrelate` here?

File: planner/core/expression_rewriter.go
Comment (Reminiscent) [inline planner/core/expression_rewriter.go:988]: Should we reset the `noDecorrelate` here?

File: planner/core/expression_rewriter.go
Comment (Reminiscent) [inline planner/core/expression_rewriter.go:1048]: Ditto

File: planner/core/expression_rewriter.go
Comment (time-and-fate) [reply planner/core/expression_rewriter.go:335]: As L331 said, first, we save the current values of `subQueryCtx` and `subQueryHintFlags`, then set them to values we want to pass to the subquery, after we're done in this function, we restore the values to the saved values.

File: planner/core/expression_rewriter.go
Comment (time-and-fate) [reply planner/core/expression_rewriter.go]: No. You can see the two examples to see golang's behavior:
https://go.dev/play/p/jg7rfs2Qc6f
https://go.dev/play/p/Cb7L2I03tvC

Comment (time-and-fate) [reply planner/core/expression_rewriter.go]: It's indeed a bit confusing, I updated the code to make it more clear.

File: planner/core/expression_rewriter.go
Comment (time-and-fate) [reply planner/core/expression_rewriter.go:321]: Updated.

File: planner/core/logical_plan_builder.go
Comment (time-and-fate) [reply planner/core/logical_plan_builder.go:3694]: Updated.
It actually doesn't matter, I added it to be consistent with the code above.

File: planner/core/expression_rewriter.go
Comment (time-and-fate) [reply planner/core/expression_rewriter.go:541]: Updated.
It actually doesn't matter, I reset it anyway to be consistent with `handleExistSubquery`.

File: planner/core/expression_rewriter.go
Comment (time-and-fate) [reply planner/core/expression_rewriter.go:990]: ditto.

File: planner/core/expression_rewriter.go
Comment (time-and-fate) [reply planner/core/expression_rewriter.go:1051]: ditto. 🤣

File: planner/core/testdata/plan_suite_in.json
Comment (time-and-fate) [reply planner/core/testdata/plan_suite_in.json:1066]: Currently, `USE_TOJA` and `NO_DECORRELATE` won't interfere with each other. `USE_TOJA` handles non-correlated `IN` subquery, and `NO_DECORRELATE` handles correlated subquery.
And `USE_TOJA` is a query level hint instead of a query block level one.

Comment (time-and-fate) [reply planner/core/testdata/plan_suite_in.json:1066]: I personally think we don't need to add such cases.
Because decorrelation-related logic happens very early in the whole optimization process, it's before any partition-related rewrites and any physical operator.
I think they won't interfere with each other. If tiflash and partitioned table related logic work well with `Apply`, there should be no other issues.

**Review by chrysan — APPROVED** [review]

**Review by Reminiscent — APPROVED** [review]

**Comment by time-and-fate** [comment]
As discussed before, let's merge this PR after release-6.3 branching.

**Pull request merged**
